### ISSUE TITLE
Add a way to detect if a cached item is presenting

### ISSUE THREAD

**Issue by pardeike** [gh123man/SwiftUI-LazyPager#13]
In my old, non-lazy code, I used .onAppear() to reset the zoom of an image in my custom item view. With LazyPager, I can't get any event that does its. Preferable before the item becomes visible when paging.

Using $index will inform my code too late.

**Comment by gh123man** [comment]
Hi, 
Each view to be lazy loaded is presented (even out of view) at the moment it is loaded. 
So you can still use `onAppear` to detect when a view is ready for presentation. You can compare it with the current displayed page to determine if it is out of view. 

Using a combination of `onAppear` and monitoring the page index (within each subview) can allow you to detect when views move in and out of view (a modified version of the [simple example](https://github.com/gh123man/SwiftUI-LazyPager/blob/master/Examples/LazyPagerExampleApp/SimpleExample.swift)):

```swift
struct SimpleExample: View {
    
    @State var data = [
        "nora1",
        "nora2",
        "nora3",
        "nora4",
        "nora5",
        "nora6",
    ]
    
    @State var show = false
    @State var page = 0
    
    var body: some View {
        LazyPager(data: data, page: $page) { element in
            Image(element)
                .resizable()
                .aspectRatio(contentMode: .fit)
                .onAppear {
                    if data.firstIndex(of: element) != page {
                        print("item: \(element) just loaded, but not visible")
                    }
                }
                .onChange(of: page) { _ in
                    if data.firstIndex(of: element) != page {
                        print("\(element) is loaded, but out of view")
                    }
                }
        }
    }
}
```

You could do a more sophisticated comparison and store the previous index to determine if a page just moved out of view. 

However it's worth pointing out that if you use `LazyPager`'s `. zoomable` function, it will automatically manage zoom, and out of view zoom resets for you.

**Comment by pardeike** [comment]
Yes, I found zoomed and am using it now. Still, let’s say I need to log whenever a user sees an item and that includes switching between B C a few times back and forth in a series A B C D E. I only get onAppear for both once and then it does not count subsequent views.

**Comment by gh123man** [comment]
Similar to my example above:

```swift
var body: some View {
        LazyPager(data: data, page: $page) { element in
            Image(element)
                .resizable()
                .aspectRatio(contentMode: .fit)
                .onChange(of: page) { _ in
                    if data.firstIndex(of: element) == page {
                        print("\(element) is displayed")
                    }
                }
        }
    }
```

by listening to `onChange` of page, and then looking it up in the dataset, you can determine when an element is in-view even if it's already been loaded.

**Comment by pardeike** [comment]
Yes, I wasn’t clear. The index switches at the end of the paging which is fine for stats but insufficient for doing things that are visual.

**Comment by gh123man** [comment]
I see. What sort of data would you need for your use case? (ex: a `float` of the current index that is updated in real time as the pager moves between pages?, something else?)

**Comment by pardeike** [comment]
Ideally, yes. I want to know where I am coming from, where I am going to and the progress. Maybe a way to know when the paging definitely will happen (peeking vs moving beyond the trigger point and releasing). But that’s maybe too much to ask for. 

Everything else seems to work fine so far. I am trying to mimic Apple’s Photos app in the way it pages, how it blends in/out its bars and how it maintains “lightbox” mode. So far I got pretty far. I’ll let you know if I get more showstoppers on the way.

**Comment by gh123man** [comment]
Ive posted a new branch: https://github.com/gh123man/SwiftUI-LazyPager/tree/absolute-content-position

> Ideally, yes. I want to know where I am coming from, where I am going to and the progress.

it exposes the absolute content position as a binding `CGFloat`

It can be used like this: 

```swift
struct SimpleExample: View {
    
    @State var data = [
        "nora1",
        "nora2",
        "nora3",
        "nora4",
        "nora5",
        "nora6",
    ]
    
    @State var show = false
    @State var contentPosition: CGFloat = 0
    
    var body: some View {
        VStack {
            LazyPager(data: data) { element in
                Image(element)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
            }
            .absoluteContentPosition($contentPosition)
            Text("\(contentPosition)")
        }
    }
}
```

> Maybe a way to know when the paging definitely will happen 

Whenever you are 50% past the current view - so you should be able to infer this from the same value. 

Can you give this branch a try and let me know if it works for you? If so - ill merge it :) 

> Everything else seems to work fine so far. I am trying to mimic Apple’s Photos app in the way it pages, how it blends in/out its bars and how it maintains “lightbox” mode. So far I got pretty far. I’ll let you know if I get more showstoppers on the way.

That's great to hear! Glad it's working for you.

**Comment by pardeike** [comment]
I just tested it. `.absoluteContentPosition($contentPosition)` works very well. While it is low level it contains just the right information to do whatever is related to paging. Well done!

**Comment by pardeike** [comment]
Question: in Apple pager, they have a "lightbox" mode. It either activates by tapping an item but also when zooming in. Right now LazyPager does not give me a way to observe the current zoom level. Could you add that too? I tried to work around it by using a weird hack but it does not work great:

```swift
@State var zoomed = false
GeometryReader { geo in
    LazyPager(data: items, page: $index) { item in
        MyItem(item)
            .onGeometryChange(for: CGSize.self) { proxy in
                return proxy.frame(in: .scrollView).size
            } action: { size in
                if item.id != index { return }
                zoomed = size.width > geo.size.width || size.height > geo.size.height
            }
    }
}
```

**Comment by pardeike** [comment]
Another thing I discovered is that if you have a custom audio player view and its among images, then the .zoomable() must be dynamic but I tried `.zoomable(min: 1, max: items[index].content.maxZoom)` and it did not make use of maxZoom = 1 for my audio player content. Maybe its applied to late? In general, having zoom control outside of the inner item is kind of annoying.

**Comment by pardeike** [comment]
Finally, I tested with 8 different items that I repeated 8000 times and this gives me huge main thread blocking whenever I page from one page to another (roughly at the 50% treshold). Its not my loading so something else is going on.

**Comment by gh123man** [comment]
> I just tested it. .absoluteContentPosition($contentPosition) works very well. While it is low level it contains just the right information to do whatever is related to paging. Well done!

Awesome! Ill merge this and cut a release soon 👍 

> Question: in Apple pager, they have a "lightbox" mode. It either activates by tapping an item but also when zooming in. Right now LazyPager does not give me a way to observe the current zoom level. Could you add that too? I tried to work around it by using a weird hack but it does not work great

Can you provide a screenshot of the lightbox? It think I'm misunderstanding what this is 😓 However yes - I can work on adding exposing the current zoom level. Let me see what I can come up with. 

> Another thing I discovered is that if you have a custom audio player view and its among images, then the .zoomable() must be dynamic but I tried .zoomable(min: 1, max: items[index].content.maxZoom) and it did not make use of maxZoom = 1 for my audio player content. Maybe its applied to late? In general, having zoom control outside of the inner item is kind of annoying.

This is a good point. There is no reason the zoom settings HAS to be global for every element. Each element gets it's own zoomable `UIScrollView` under the hood. I'll try to come up with an alternative API to expose per element zoom controls.

**Comment by pardeike** [comment]
Thanks for the great feedback. Also good to see that this project is alive and kicking. I am going to use it in the government organization I am currently working in.

The "lightbox" is more a cumulative term for what Apple does in its Photos app. It toggles between distraction free mode (black background, no ux) and "normal" background plus some UX for the given item. On an image the normal mode moves in black bars from top and bottom which contain header elements and footer buttons plus a carusell to move around with. With Video you get some extra playback controls.

This means my use case requires some tricky gesture coordination. I have the normal swiping that LazyPager has, I have to handle taps to toggle the lightbox mode, I have the zoom functionality (that also implicitly turns on lightbox mode when zooming in and turns it of when it wasn't tap-on before, so complicated state logic). Finally, paging should work when already zoomed in. Most of that I got working with LazyPager but zooming only with the previously described hack (which works "ok" but is ugly). Finally, when zooming below 80% or os, iOS exits from paging and switches to a grid. I have not yet implemented that and I think our app does not need that. I recommend checking out standard Photos app in 18.2 to get a feeling of the rich interactions that users are used to.

**Comment by gh123man** [comment]
> Thanks for the great feedback. Also good to see that this project is alive and kicking. I am going to use it in the government organization I am currently working in.

Thank you for the feedback and feature suggestions! My pleasure to accommodate. 

> The "lightbox" is more a cumulative term for what Apple does in its Photos app. It toggles between distraction free mode (black background, no ux) and "normal" background plus some UX for the given item. On an image the normal mode moves in black bars from top and bottom which contain header elements and footer buttons plus a carusell to move around with. With Video you get some extra playback controls.

Thank you! I understand now 👍

Ive made some more changes, and added two new API's for you to try. 

1. Per element zoom controls. Alongside the existing `.zoomable` function, there is a new: `func zoomable(onElement: @escaping (Element) -> ZoomConfig) -> LazyPager` which provides an element to you, and you can return the zoom config for this element. This allows you to control the zoom settings on a per element basis. This callback is called every time the view state is updated, which is hopefully frequent enough for you to update the zoom settings dynamically. Note - if you want to derive zoom settings from view geometry, you will have to store it back on the `element` so it can be handled here. 

2. I added a `onZoom(_ onZoomHandler: @escaping (Element, CGFloat) -> ()) -> LazyPager` function that provides you with an element, and the current zoom level (between `min` and `max`). This is called during a zoom so the value is continuously updated while a zoom is happening. 

Here is a complete example of both API's in use, and these changes are published on this branch: https://github.com/gh123man/SwiftUI-LazyPager/tree/custom-zoom

```swift
struct SimpleExample: View {
    
    @State var data: [(String, CGFloat, CGFloat)] = [
        ("nora1", 1, 2),
        ("nora2", 1, 3),
        ("nora3", 1, 4),
        ("nora4", 1, 1), // Disabled
        ("nora5", 1, 6),
        ("nora6", 1, 7),
    ]
    
    @State var show = false
    @State var contentPosition: CGFloat = 0
    
    var body: some View {
        VStack {
            LazyPager(data: data) { element in
                Image(element.0)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
            }
            .absoluteContentPosition($contentPosition)
            .onZoom { element, zoomLevel in
                print(element.0, " is at zoom level: ", zoomLevel)
            }
            .zoomable { element in
                return .custom(min: element.1, max: element.2, doubleTap: .scale(0.5))
            }
            Text("\(contentPosition)")
        }
    }
}
```

Please let me know how this works, and if it helps your use case. 
Note - I haven't tested this super thoroughly, so if you find a bug please let me know and ill fix it :)

**Comment by pardeike** [comment]
So far, works great. The value of `doubleTap` (.scale) is confusing though.

Also, I found myself doing

```swift
var index = 0
LazyPager(data: items, page: pageBinding) {
    // ...
    let _ = (index += 1)
}
```

to get an index. While its fine it would be great to have that build-in. Since I am using any Foo I cannot easily make items in `items` Equatable or else I would look up their position in the array.

**Comment by pardeike** [comment]
I reworked my design so the extra index isn't super important. Imho you can make a release with the current features as they are.

Just for reference, here is my current demo code main view:

```swift
public var body: some View {
    ZStack {
        (isInLightboxMode ? Color.black : Color(UIColor.systemBackground))
        GeometryReader { _ in
            LazyPager(data: items, page: pageBinding) {
                AnyView(contentView($0))
            }
            .settings { $0.preloadAmount = 1 }
            .absoluteContentPosition($absPosition)
            .onZoom { _, zoomLevel in
                if absPosition != CGFloat(Int(absPosition)) { return }
                if zoomLevel > 1 {
                    withAnimation {
                        lightbox = true
                    }
                } else if lightboxLocked == false {
                    withAnimation {
                        lightbox = false
                    }
                }
            }
            .zoomable {
                getItem($0)?.zoomConfig() ?? .disabled
            }
            .onTap {
                toggleLightbox()
            }
            .onChange(of: index) {
                getItem($0)?.visibilityChanged(visible: false)
                getItem($1)?.visibilityChanged(visible: true)
            }
            .onAppear {
                if items.isEmpty { return }
                getItem(items[0])?.visibilityChanged(visible: true)
            }
        }
        if !items.isEmpty && !isInLightboxMode && showToolbar {
            overlayView?(index, items[index])
            AnyView(itemOverlayView(items[index]))
        }
    }
}
```

The only ugly thing that I can spot is that since a lot of view modifiers are outside the loop, I have to use items[index] which is kind of weird. But not a big deal.

**Comment by gh123man** [comment]
> So far, works great. The value of doubleTap (.scale) is confusing though.

Ill try to make this more clear in the readme. `.scale` takes a CGFloat (between `0` and `1`) that represents the amount between `min` and `max` you want to zoom. So `.scale(0.5)` is 50% between `1` and `10`. 

>I reworked my design so the extra index isn't super important. Imho you can make a release with the current features as they are.

Awesome, Ill release this today! 

Thanks for sharing your code, great to see how this is being used - and it helps me design better API's 👍 
And thanks again for all the feedback and input!

**Comment by gh123man** [comment]
Published in `1.1.6`
Please let me know if you have any other issues. 
If this answers your questions for now, i'm going to close this issue :)

**Comment by pardeike** [comment]
Nice work. Love it
